### DISCUSSION TITLE
How can I parse `PackageReference` from many `ItemGroup`?

### DISCUSSION THREAD

**Discussion by 6a6f6a6f** [dotnet/msbuild#7523] [Q&A]
I'm using `Microsoft.Build` on version `17.1.0` as a library inside a .NET 6 simple application. My main goal is to have the attributes of `PackageReference` (`Include` and `Version`) starting at `ProjectRootElement.Open(path)`, but when I loop thou the `ItemGroups`, I'm not able to find any other attribute than `Version`:

```csharp
project.ItemGroups
    .SelectMany(rootElement => rootElement.Items)
    .Where(itemElement => itemElement.ElementName == "PackageReference")
    .ToList()
    .ForEach(item =>
    {
        var metadata = item.Metadata.ToList();
        metadata.ForEach(d =>
        {
            Console.WriteLine($"{d.Name}: {d.Value}");
        });
    });
```

If these questions do not fit here, please bring me some direction.

Many thanks for considering my request.

Information about my runtime:
- .NET SDK: 6.0.201 (commit `ef40e6aa06`);
- OS Name: Mac OS X;
- OS Version:  12.2;
- OS Platform: Darwin;
- RID: osx.12-arm64.

**Comment by rainersigwald** [comment]
Are you wanting to do this _as XML_, or _as MSBuild objects_? The latter is preferred, but requires you to think more about the actual item type. Specifically, the `Include` is _expressed in XML as an attribute_, but becomes something more fundamental to the MSBuild type--it's not metadata because it's special and it's available in [`ProjectItemElement.Include`](https://docs.microsoft.com/dotnet/api/microsoft.build.construction.projectitemelement.include?view=msbuild-17-netcore#microsoft-build-construction-projectitemelement-include). `Exclude` and `Remove` are also special.

What specific problem are you trying to solve? There may be a better way to go about it.

**Reply by 6a6f6a6f** [reply]
Thank you for moving to the right place @rainersigwald. The problem I'm trying to solve is to extract all references and dependencies of a project, including the version.

**Reply by rainersigwald** [reply]
In that case, you should think about _evaluating_ the project using [`Microsoft.Build.Evaluation.Project.GetItems`](https://docs.microsoft.com/dotnet/api/microsoft.build.evaluation.project.getitems) rather than using the `Microsoft.Build.Construction` API. That's because the Construction API looks only at _the project file_, but a user might define references in an imported file or use something like Central Package Versioning to apply versions in a place other than "directly in the project file".
